app/src/main/java/com/example/habitsmasher/FollowUserDialog.java

```diff
@@ -0,0 +1,160 @@
+
+        // Attach UI elements
+        _userToFollow = view.findViewById(R.id.user_search_text);
+        Button cancelButton = view.findViewById(R.id.cancel_user_follow);
+        Button followButton = view.findViewById(R.id.follow_user_button);
+
+        // cancel button logic
+        cancelButton.setOnClickListener(new ClickListenerForCancel(followUserDialog, TAG));
+
+        // follow button logic
+        followButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                Log.d(TAG, "Follow");
+
+                String usernameInput = _userToFollow.getText().toString().trim();
+
+                if (usernameInput.isEmpty()) {
+                    displayErrorMessage(EMPTY_USERNAME_ERROR);
+                    return;
+                }
+
+                CollectionReference usersCollectionRef = _db.collection(USERS_COLLECTION_PATH);
+                usersCollectionRef.get().addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
+                    @Override
+                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
+                        if (task.isSuccessful()) {
+                            for(QueryDocumentSnapshot document: task.getResult()){
+                                if (document.exists()) {
+                                    if (document.get(USERNAME_FIELD).toString().equals(usernameInput)){
+                                        Log.d(TAG, "Username exists");
+
+                                        // if username exists, update follower/following counts
+                                        String currentUserId = getCurrentUserId();
+
+                                        User userToFollow = document.toObject(User.class);
+                                        userToFollow.addNewFollower(currentUserId);
+
+                                        // add new follower to the user to follow
+                                        addNewFollowerForUserInDatabase(userToFollow.getId(),
+                                                                        currentUserId);
+
+                                        // add to new followed user to following of current user
+                                        addUserToFollowingForUserInDatabase(currentUserId,
+                                                                            userToFollow.getId());
+
+                                        showFollowSuccessMessage();
+
+                                        followUserDialog.dismiss();
+
+                                        return;
+                                    }
+                                }
+                            }
+                            // if username does not exist, throw error message on the front-end
+                            Log.d(TAG, "Username does not exist");
+
+                            displayErrorMessage(INVALID_USERNAME_ERROR);
```

Comment:
I couldn't find a better way to do this. What I'm trying to do here is **check if a specific username exists** in the database. You have to iterate through everything and manually check apparently.

To explain this block of code:
* Use the **for loop** to iterate through each user in the "Users" collection
* Check if the username of the current user in the iteration equals what the username entered
* If yes... Add the current user as a **follower** of the user to follow and add the newly followed user to the **following** list of the current user
* If no, this means that the username does not exist, therefore it is invalid. Show an error in the front end!